### PR TITLE
Misc inflate cleanup

### PR DESCRIPTION
**Pull request by Dead2**
- Stop updating 'dmax' and 'sane' and remove the variables from 'inflate_state' unless the checks are compiled in.
- Add a variable 'wbufsize' to track the real size of the window buffer, including padding. This allows chunkset code to spill garbage writes into the padding area if it is available. It is not available for inflateBack unfortunately.
- Reorder the 'inflate_state' struct to improve cache-locality of the variables needed by inffast (reduced from 6 to 1 cacheline)
  Also fills in some unnecessary holes.
  
On x86-64, default settings, inflate_state also shrinks from 9216 to 9152 bytes and now ends cleanly on a cacheline boundary.

### CODE REVIEW

**Comment by Dead2** [comment]
### x86-64 (i9-9900K) 2.2.2
 ```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.185%      0.104/0.109/0.111/0.002        0.030/0.035/0.038/0.002        8,526,745
 2     43.871%      0.198/0.203/0.204/0.001        0.032/0.036/0.038/0.002        6,903,702
 3     42.390%      0.247/0.250/0.253/0.002        0.032/0.035/0.036/0.001        6,670,664
 4     41.644%      0.275/0.280/0.282/0.002        0.031/0.034/0.035/0.001        6,553,205
 5     41.215%      0.303/0.308/0.310/0.002        0.031/0.034/0.035/0.001        6,485,659
 6     41.032%      0.348/0.355/0.359/0.003        0.029/0.033/0.035/0.002        6,456,912
 7     40.778%      0.496/0.505/0.509/0.003        0.031/0.033/0.035/0.001        6,416,941
 8     40.704%      0.613/0.619/0.624/0.003        0.032/0.033/0.035/0.001        6,405,249
 9     40.409%      0.906/0.911/0.915/0.003        0.030/0.033/0.035/0.001        6,358,951

 avg1  42.914%                        0.393                          0.034
 tot                                106.189                          9.189       60,778,028

   text    data     bss     dec     hex filename
 130774    1312       8  132094   203fe libz-ng.so.2

          5,521.85 msec task-clock:u                     #    1.000 CPUs utilized               ( +-  0.03% )
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
               144      page-faults:u                    #   26.078 /sec                        ( +-  0.39% )
    19,093,810,958      cycles:u                         #    3.458 GHz                         ( +-  0.03% )  (62.49%)
    35,726,573,927      instructions:u                   #    1.87  insn per cycle              ( +-  0.01% )  (74.99%)
     3,758,372,000      branches:u                       #  680.636 M/sec                       ( +-  0.02% )  (74.99%)
       174,901,372      branch-misses:u                  #    4.65% of all branches             ( +-  0.03% )  (74.99%)
     6,865,961,607      L1-dcache-loads:u                #    1.243 G/sec                       ( +-  0.02% )  (74.99%)
       233,340,453      L1-dcache-load-misses:u          #    3.40% of all L1-dcache accesses   ( +-  0.07% )  (75.01%)
        18,378,169      LLC-loads:u                      #    3.328 M/sec                       ( +-  1.58% )  (50.01%)
               144      LLC-load-misses:u                #    0.00% of all LL-cache accesses    ( +- 20.28% )  (50.00%)
```

### x86-64 Inflate cleanup PR
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.185%      0.105/0.109/0.111/0.002        0.030/0.036/0.038/0.002        8,526,745
 2     43.871%      0.194/0.197/0.198/0.001        0.033/0.036/0.037/0.001        6,903,702
 3     42.390%      0.240/0.244/0.246/0.002        0.030/0.034/0.036/0.002        6,670,664
 4     41.644%      0.270/0.275/0.277/0.002        0.029/0.033/0.035/0.001        6,553,205
 5     41.215%      0.293/0.299/0.302/0.002        0.031/0.034/0.035/0.001        6,485,659
 6     41.032%      0.341/0.348/0.352/0.003        0.029/0.033/0.035/0.002        6,456,912
 7     40.778%      0.491/0.501/0.507/0.004        0.031/0.033/0.035/0.001        6,416,941
 8     40.704%      0.609/0.616/0.621/0.003        0.031/0.033/0.035/0.001        6,405,249
 9     40.409%      0.904/0.911/0.917/0.003        0.029/0.033/0.034/0.001        6,358,951

 avg1  42.914%                        0.389                          0.034
 tot                                104.973                          9.172       60,778,028

   text    data     bss     dec     hex filename
 130102    1312       8  131422   2015e libz-ng.so.2

          5,518.07 msec task-clock:u                     #    1.000 CPUs utilized               ( +-  0.02% )
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
               144      page-faults:u                    #   26.096 /sec                        ( +-  0.37% )
    19,068,549,752      cycles:u                         #    3.456 GHz                         ( +-  0.03% )  (62.47%)
    35,717,171,689      instructions:u                   #    1.87  insn per cycle              ( +-  0.01% )  (74.98%)
     3,758,323,631      branches:u                       #  681.094 M/sec                       ( +-  0.02% )  (74.98%)
       175,939,466      branch-misses:u                  #    4.68% of all branches             ( +-  0.06% )  (75.00%)
     6,864,926,063      L1-dcache-loads:u                #    1.244 G/sec                       ( +-  0.03% )  (75.02%)
       232,739,294      L1-dcache-load-misses:u          #    3.39% of all L1-dcache accesses   ( +-  0.07% )  (75.00%)
        18,475,946      LLC-loads:u                      #    3.348 M/sec                       ( +-  0.95% )  (50.00%)
               124      LLC-load-misses:u                #    0.00% of all LL-cache accesses    ( +- 24.54% )  (49.99%)
```

### Rpi5b 2.2.2
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.000/0.000/0.000/0.000        0.000/0.000/0.003/0.001       15,737,543
 1     54.185%      0.096/0.109/0.112/0.004        0.028/0.041/0.047/0.005        8,526,745
 2     43.871%      0.184/0.192/0.195/0.003        0.027/0.040/0.044/0.004        6,903,702
 3     42.390%      0.216/0.230/0.236/0.004        0.031/0.040/0.045/0.003        6,670,664
 4     41.644%      0.250/0.262/0.266/0.003        0.024/0.034/0.040/0.004        6,553,205
 5     41.215%      0.276/0.286/0.290/0.004        0.020/0.036/0.040/0.005        6,485,659
 6     41.032%      0.323/0.336/0.340/0.004        0.024/0.036/0.040/0.004        6,456,912
 7     40.778%      0.428/0.446/0.451/0.005        0.028/0.037/0.040/0.003        6,416,941
 8     40.704%      0.562/0.568/0.573/0.002        0.020/0.037/0.040/0.004        6,405,249
 9     40.409%      0.679/0.684/0.688/0.002        0.027/0.037/0.042/0.003        6,358,951

 avg1  48.624%                        0.311                          0.034
 avg2  54.026%                        0.346                          0.038
 tot                                124.488                         13.537       76,515,571

   text    data     bss     dec     hex filename
 109760    1504       8  111272   1b2a8 libz-ng.so.2
```
 
### Rpi5b Inflate cleanup PR
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.000/0.000/0.000/0.000        0.000/0.000/0.000/0.000       15,737,543
 1     54.185%      0.096/0.109/0.112/0.004        0.031/0.040/0.043/0.003        8,526,745
 2     43.871%      0.184/0.192/0.197/0.004        0.025/0.039/0.042/0.004        6,903,702
 3     42.390%      0.226/0.232/0.234/0.002        0.032/0.039/0.044/0.003        6,670,664
 4     41.644%      0.254/0.262/0.266/0.003        0.021/0.035/0.039/0.004        6,553,205
 5     41.215%      0.281/0.288/0.293/0.003        0.030/0.037/0.042/0.003        6,485,659
 6     41.032%      0.319/0.338/0.342/0.004        0.031/0.036/0.039/0.003        6,456,912
 7     40.778%      0.436/0.449/0.452/0.004        0.023/0.034/0.039/0.004        6,416,941
 8     40.704%      0.558/0.570/0.574/0.003        0.020/0.035/0.039/0.005        6,405,249
 9     40.409%      0.681/0.689/0.692/0.003        0.025/0.034/0.038/0.003        6,358,951

 avg1  48.624%                        0.313                          0.033
 avg2  54.026%                        0.348                          0.037
 tot                                125.146                         13.153       76,515,571

   text    data     bss     dec     hex filename
 109496    1504       8  111008   1b1a0 libz-ng.so.2
```

TLDR;
x86-64 deflatebench: 0.15% speedup
x86-64 perf: 0.069% speedup
Codesize reduced by 672 bytes

Aarch64 deflatebench: 2.83% speedup (Probably just hit a sweet-spot and not reliable)
Codesize reduced by 264 bytes

All in all it seems likely to be a minor speedup along with a slight reduction of codesize.
It should also make it easier to gain more speed from chunkset code improvements later.

**Comment by Dead2** [comment]
PS: Betatesting coderabbitai, so far it does not seem too intelligent when it comes to these changes 😆

**Comment by Dead2** [comment]
MSVC CI is failing due to warning C4324 (struct padded due to alignment). I don't really think that warning (especially as an error) provides us any benefit at all. The fact that the end of a struct is padded up to the next cacheline boundary is pretty much a non-issue.

@mtl1979 Perhaps we should just disable that warning entirely? It would end up failing if someone compiles with optional features anyway, and it also makes it harder to maintain since not everyone has access to MSVC.

File: inflate.h [hunk 1]
```diff
@@ -97,59 +97,66 @@ typedef struct inflate_allocs_s {
...
+    uint32_t chunksize;         /* size of memory copying chunk */
 
         /* bit accumulator */
     uint32_t hold;              /* input bit accumulator */
     unsigned bits;              /* number of bits in "in" */
+        /* fixed and dynamic code tables */
+    unsigned lenbits;           /* index bits for lencode */
+    code const *lencode;        /* starting table for length/literal codes */
+    code const *distcode;       /* starting table for distance codes */
+    unsigned distbits;          /* index bits for distcode */
```
[the remark this answers is not in the document]
Comment (Dead2) [reply inflate.h:130]: These fields are not new, they have just been moved 5 lines up.

File: inflate.c [hunk 2]
```diff
@@ -73,13 +73,17 @@
...
-    state->dmax = 32768U;
     state->head = NULL;
     state->hold = 0;
     state->bits = 0;
     state->lencode = state->distcode = state->next = state->codes;
-    state->sane = 1;
     state->back = -1;
+#ifdef INFLATE_STRICT
+    state->dmax = 32768U;
+#endif
```
[the remark this answers is not in the document]
Comment (Dead2) [reply inflate.c:83]: state->dmax has been completely removed if INFLATE_STRICT is not set.

File: inflate.c [hunk 3]
```diff
@@ -73,13 +73,17 @@
...
+#ifdef INFLATE_ALLOW_INVALID_DISTANCE_TOOFAR_ARRR
+    state->sane = 1;
+#endif
```
[the remark this answers is not in the document]
Comment (Dead2) [reply inflate.c:86]: state->sane has been completely removed if INFLATE_ALLOW_INVALID_DISTANCE_TOOFAR_ARRR is not set.

File: inflate.c [hunk 4]
```diff
@@ -539,7 +544,9 @@
                 SET_BAD("invalid window size");
                 break;
             }
+#ifdef INFLATE_STRICT
             state->dmax = 1U << len;
+#endif
```
[the remark this answers is not in the document]
Comment (Dead2) [reply inflate.c:549]: state->dmax has been completely removed if INFLATE_STRICT is not set.

**Comment by mtl1979** [comment]
> @mtl1979 Perhaps we should just disable that warning entirely? It would end up failing if someone compiles with optional features anyway, and it also makes it harder to maintain since not everyone has access to MSVC.

Some warnings are safe to disable, disabling others can cause crashes or poor performance... We have line in CMakeLists.txt that disables useless warnings.

**Comment by Dead2** [comment]
> Some warnings are safe to disable, disabling others can cause crashes or poor performance... We have line in CMakeLists.txt that disables useless warnings.

Yeah, I meant to add a commit that disables that warning here, but instead I managed to push it directly into Develop. Can't force-push without it, so I think we'll just keep it rather than revert it and add it back by PR.

File: infback.c [hunk 5]
```diff
@@ -53,14 +53,19 @@ int32_t ZNG_CONDEXPORT PREFIX(inflateBackInit)(PREFIX3(stream) *strm, int32_t wi
     Tracev((stderr, "inflate: allocated\n"));
 
     strm->state = (struct internal_state *)state;
-    state->dmax = 32768U;
     state->wbits = (unsigned int)windowBits;
     state->wsize = 1U << windowBits;
+    state->wsize_real = 1U << windowBits;
```
Comment (nmoinvaz) [inline infback.c]: nit: optional naming if you want: `wbufsize`?

**Review by nmoinvaz — APPROVED** [review]
Nice improvements and nice code size reduction.

**Review by KungFuJesus — APPROVED** [review]
LGTM

File: infback.c (hunk 5 above)
Comment (Dead2) [reply infback.c]: That is a better name, thanks 👍

File: inflate.h [hunk 6]
```diff
@@ -97,59 +97,66 @@ typedef struct inflate_allocs_s {
...
+    PREFIX3(stream) *strm;      /* pointer back to this zlib stream */
     inflate_mode mode;          /* current inflate mode */
     int last;                   /* true if processing last block */
     int wrap;                   /* bit 0 true for zlib, bit 1 true for gzip,
                                    bit 2 true to validate check value */
     int havedict;               /* true if dictionary provided */
     int flags;                  /* gzip header method and flags, 0 if zlib, or
                                    -1 if raw or no header yet */
-    unsigned dmax;              /* zlib header max distance (INFLATE_STRICT) */
+    unsigned was;               /* initial length of match, for inflateMark */
```
[the remark this answers is not in the document]
Comment (Dead2) [reply inflate.h:108]: state->was is used to implement the `inflateMark()` function in `inflate.c`.

File: infback.c [hunk 7]
```diff
@@ -53,14 +53,19 @@ int32_t ZNG_CONDEXPORT PREFIX(inflateBackInit)(PREFIX3(stream) *strm, int32_t wi
     Tracev((stderr, "inflate: allocated\n"));
 
     strm->state = (struct internal_state *)state;
-    state->dmax = 32768U;
     state->wbits = (unsigned int)windowBits;
     state->wsize = 1U << windowBits;
+    state->wbufsize = 1U << windowBits;
```
[the remark this answers is not in the document]
Comment (Dead2) [reply infback.c:58]: The PR description had already been updated when you left this feedback.
